### PR TITLE
Add Test for CL_KERNEL_ATTRIBUTES (#832)

### PR DESCRIPTION
**Pull request by chemis01**
This test generates dummy kernels which have different
combinations of the following attributes:

    * vec_type_hint
    * work_group_size_hint
    * reqd_work_group_size

It then gets the attributes by using clGetKernelInfo
and validates that the attributes returned are correct
by matching the attributes which were used to generate
the kernel are present in the returned string from
clGetKernelInfo.

Additionally, this PR adds some functionality to the Harness to make it easier and more intuitive to know if a device supports the Half or Double types. 

Signed-off-by: Chetankumar Mistry <chetan.mistry@arm.com>

### CODE REVIEW

**Comment by jrprice** [comment]
Not a thorough review, but please note that 64-bit integer types (`long`/`ulong`) are optional for embedded profile, so these should not be tested unconditionally. They should be guarded with checks similar to `double` and `half`, e.g. if the device is embedded profile only run them if `cles_khr_int64` is supported.

File: test_conformance/api/test_kernel_attributes.cpp
Comment (StuartDBrady) [inline test_conformance/api/test_kernel_attributes.cpp:300]: This output could be made a little clearer.  Would suggest `"ERROR: did not find expected attribute: '%s'\n"`, with the full attribute list printed just once if a failure is found for the permutation, with a message such as `"Attribute list reported is: '%s'\n"`.

I would note that this doesn't raise an error if extra attributes were reported that were not in the permutation. Should it?

Comment (StuartDBrady) [reply test_conformance/api/test_kernel_attributes.cpp]: The most recent changes look good to me. I see now that the `CL_KERNEL_ATTRIBUTES` info is not trivial to parse. I think it would be good to check that there are no extraneous attributes, but this could be added some other time.

Would be good to have independent review of the rest of this PR.

**Review by StuartDBrady — COMMENTED** [review]
This looks good to me but it would be good to have some independent review.

**Comment by chemis01** [comment]
Hi all,
Are there any issues with this PR? I'd be happy to make some changes so that this can be merged.
Thanks

File: test_conformance/api/test_kernel_attributes.cpp
Comment (jrprice) [inline test_conformance/api/test_kernel_attributes.cpp:26]: I think this should be removed.

File: test_conformance/api/test_kernel_attributes.cpp
Comment (jrprice) [inline test_conformance/api/test_kernel_attributes.cpp:142]: Very nitty style point, but I think we should try to consistently use C++ comments in most places (`//`), as we're generally moving the codebase to be more canonically C++.

File: test_conformance/api/test_kernel_attributes.cpp
Comment (jrprice) [inline test_conformance/api/test_kernel_attributes.cpp:282]: I think the checks here should use `test_error` instead of `ASSERT_SUCCESS`. The latter will throw an exception and abort the run, which doesn't play nicely when not running this test in isolation.

**Review by jrprice — CHANGES_REQUESTED** [review]
Looks mostly fine, just a couple of nits and one specific thing I think should be changed.

File: test_conformance/api/test_kernel_attributes.cpp
Comment (chemis01) [reply test_conformance/api/test_kernel_attributes.cpp:26]: Ok.
Keeping parseParameters.h as it is needed for determining the compilation mode.

Comment (jrprice) [reply test_conformance/api/test_kernel_attributes.cpp]: I was just talking the unused macro definition (GitHub added the extra context automatically). The `kernelHelpers.h` include could be left in for IWYU reasons since there are functions from it used here, but I'm not going to make you change this again before merging.

**Review by jrprice — DISMISSED** [review]
LGTM, thanks for the updates.

As an aside, I'd find it much easier/quicker to review updates to PRs if they were pushed as new commits rather than with amend+force-push. When you rewrite history like this GitHub doesn't let you see what changed since the last time you reviewed, which means you have to review the whole file again. We always squash the history when merging PRs so the extra commits don't matter.

**Review by bashbaug — COMMENTED** [review]
We're discussing this test internally.  Could you please hold off merging it for a little longer?  Thanks!

**Comment by kpet** [comment]
@bashbaug Any update?

**Comment by bashbaug** [comment]
Sorry for the very slow reply - somehow I missed the "mention" nag above.

We're currently failing this test, though the reason why is somewhat interesting.  As background, we compile through SPIR-V by default even in the compile-from-source path, so we rely on SPIR-V to carry through all attributes.

At the SPIR-V level, we have:

VecTypeHint
A hint to the compiler, which indicates that most operations used in the entry point are explicitly vectorized using a particular vector type. The 16 high-order bits of the Vector Type operand specify the number of components of the vector. The 16 low-order bits of the Vector Type operand specify the data type of the vector.

These are the legal data type values:
0 represents an 8-bit integer value.
1 represents a 16-bit integer value.
2 represents a 32-bit integer value.
3 represents a 64-bit integer value.
4 represents a 16-bit float value.
5 represents a 32-bit float value.
6 represents a 64-bit float value.

Note that SPIR-V does not distinguish between signed and unsigned integer values, so we're losing the signedness of the `vec_type_hint`, and either failing the test for a signed attribute or an unsigned attribute since we have to pick one arbitrarily.

If we want to make short-term forward progress with this PR I think we have several options:

1. Enable the `work_group_size_hint` and `reqd_work_group_size` tests but keep `vec_type_hint` test disabled.
2. Enable all three tests and either pass signed or unsigned integer types for the `vec_type_hint` or warn on this behavior, but don't fail the test.

Longer-term we need to decide what we want to do for this attribute and/or the query and either fix the OpenCL spec, find a way to carry the signedness through SPIR-V, or we'll need to find some other way to carry the attributes through our implementation.

**Comment by bashbaug** [comment]
Hi, we're still failing the new test for the reason described above:

```
$ ./test_conformance/api/test_api kernel_attributes
 Initializing random seed to 0.
Requesting Default device based on command line for platform index 0 and device index 0
Compute Device Name = Intel(R) Graphics [0x5916], Compute Device Vendor = Intel(R) Corporation, Compute Device Version = OpenCL 3.0 NEO , CL C Version = OpenCL C 3.0 
Device latest conformance version passed: v2020-11-23-00
Supports single precision denormals: YES
sizeof( void*) = 8  (host)
sizeof( void*) = 8  (device)
kernel_attributes...
ERROR: did not find expected attribute: 'vec_type_hint(char2)'
Attribute string reported as: vec_type_hint(uchar2)
...
```

**Comment by kpet** [comment]
@bashbaug Interesting. We've removed the signed cases to get forward progress on this PR (@chemis01 You may want to leave a TODO/FIXME in the code pointing at a new issue in this project). Let us know if you're now passing the updated tests.

File: test_conformance/api/test_kernel_attributes.cpp
Comment (bashbaug) [inline test_conformance/api/test_kernel_attributes.cpp:55]: This probably does not matter in the grand scheme of things, but I do observe that these permutations are stored in global variables and never freed or cleared, so all of the permutations will persist beyond the execution of this test.  It would be nicer to store them in local variables so the destructors will run when the test completes, or to clear their contents before the test returns pass or fail, but "leaking" the permutations is probably fine for most systems with a reasonable amount of memory.

**Review by bashbaug — APPROVED** [review]
Thanks, with the latest changes these tests pass on our implementation.

I left one comment, but please consider it a nice-to-have and not a must-have, and I'm fine with these changes being merged as-is.

**Comment by kpet** [comment]
Remaining comments to be addressed under https://github.com/KhronosGroup/OpenCL-CTS/issues/1248

**Pull request merged**
